schema: Enable deprecations in cc_users_groups

aciba90 opened the pull request:
## Proposed Commit Message
<!-- Include a proposed commit message because all PRs are squash merged -->

```
schema: Enable deprecations in cc_users_groups

Extend doc of schemas with multiple sub-schemas including.
```

## Additional Context
<!-- If relevant -->
SC-1173
SC-1175

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

blackboxsw commented on cloudinit/config/cc_users_groups.py:
```diff
@@ -141,7 +141,7 @@
         user:
           name: mynewdefault
-          sudo: false
+          sudo: null
```
Wow this is a bug I introduced in a63f45f718. Cloud-init doesn't handle this deprecation or warn about it ... and it gets a traceback when launching with this suggested content.

```bash
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/cloudinit/config/modules.py", line 246, in _run_modules
    ran, _r = cc.run(
  File "/usr/lib/python3/dist-packages/cloudinit/cloud.py", line 67, in run
    return self._runners.run(name, functor, args, freq, clear_on_fail)
  File "/usr/lib/python3/dist-packages/cloudinit/helpers.py", line 185, in run
    results = functor(*args)
  File "/usr/lib/python3/dist-packages/cloudinit/config/cc_users_groups.py", line 206, in handle
    cloud.distro.create_user(user, **config)
  File "/usr/lib/python3/dist-packages/cloudinit/distros/__init__.py", line 655, in create_user
    self.write_sudo_rules(name, kwargs["sudo"])
  File "/usr/lib/python3/dist-packages/cloudinit/distros/__init__.py", line 814, in write_sudo_rules
    raise TypeError(msg % (type_utils.obj_name(rules)))
TypeError: Can not create sudoers rule addition with type 'NoneType'
```



Here's a logic fix that will also emit warnings on deprecated value:
```diff
diff --git a/cloudinit/distros/__init__.py b/cloudinit/distros/__init__.py
index ffa41093f..f2c744440 100644
--- a/cloudinit/distros/__init__.py
+++ b/cloudinit/distros/__init__.py
@@ -651,8 +651,15 @@ class Distro(persistence.CloudInitPickleMixin, metaclass=abc.ABCMeta):
             self.lock_passwd(name)
 
         # Configure sudo access
-        if "sudo" in kwargs and kwargs["sudo"] is not False:
-            self.write_sudo_rules(name, kwargs["sudo"])
+        if "sudo" in kwargs:
+            if kwargs["sudo"]:
+                self.write_sudo_rules(name, kwargs["sudo"])
+            elif kwargs["sudo"] is False:
+                LOG.warning(
+                    "DEPRECATED: The user %s has a 'sudo' config value of"
+                    " 'false' which is deprecated and will be removed in a"
+                    " future version. Use 'null' instead."
+                )
 
         # Import SSH keys
         if "ssh_authorized_keys" in kwargs:
```

blackboxsw replied:
I pushed a commit c60e194b931ac88abd48c8887f6e77c0db185a64 on top of this PR for ease of your review. If undesireable, feel free to pop it off and refactor what you want.

blackboxsw reviewed (commented):
Ok full diff of some review comments, I'm going to make a separate PR for this as it's unrelated to the current PR:
 - resolve UNDEFINED on subtypes.
 - avoid doc duplication of where top-level type defined and enum defined
 - avoid doc duplication where top-level type defined and allOf/anyOff also defines individual types
 - fix schema bug for cc_users_groups sude: null handling
 

```diff
diff --git a/cloudinit/config/schema.py b/cloudinit/config/schema.py
index da86298dd..8e2131bbe 100644
--- a/cloudinit/config/schema.py
+++ b/cloudinit/config/schema.py
@@ -64,6 +64,7 @@ SCHEMA_LIST_ITEM_TMPL = (
 SCHEMA_EXAMPLES_HEADER = "**Examples**::\n\n"
 SCHEMA_EXAMPLES_SPACER_TEMPLATE = "\n    # --- Example{0} ---"
 DEPRECATED_KEY = "deprecated"
+DEPRECATED_PREFIX = "DEPRECATED. "
 
 
 # type-annotate only if type-checking.
@@ -164,10 +165,10 @@ def is_schema_byte_string(checker, instance):
 
 
 def _add_deprecation_msg(description: Optional[str] = None):
-    msg = "DEPRECATED."
+    msg = DEPRECATED_PREFIX
     if description:
-        msg += f" {description}"
-    return msg
+        msg += f"{description}"
+    return msg.strip()
 
 
 def _validator_deprecated(
@@ -725,11 +726,19 @@ def _flatten(xs):
 
 
 def _collect_subschema_types(property_dict: dict, multi_key: str) -> List[str]:
-    property_types = [
-        subschema["type"]
-        for subschema in property_dict.get(multi_key, {})
-        if subschema.get("type") and not subschema.get("deprecated", False)
-    ]
+    property_types = []
+    for subschema in property_dict.get(multi_key, {}):
+        if subschema.get(DEPRECATED_KEY):  # don't document deprecated types
+            continue
+        if subschema.get("enum"):
+            property_types.extend(
+                [
+                    f"``{_YAML_MAP.get(enum_value, enum_value)}``"
+                    for enum_value in subschema.get("enum", [])
+                ]
+            )
+        elif subschema.get("type"):
+            property_types.append(subschema["type"])
     return list(_flatten(property_types))
 
 
@@ -761,7 +770,10 @@ def _get_property_type(property_dict: dict, defs: dict) -> str:
         sub_property_types = [sub_property_types]
     # Collect each item type
     for sub_item in chain(items.get("oneOf", {}), items.get("anyOf", {})):
-        sub_property_types.append(_get_property_type(sub_item, defs))
+        sub_property_type = _get_property_type(sub_item, defs)
+        if sub_property_type != "UNDEFINED":
+            # anyOf and oneOf can inherit parent type may have UNDEFINED type
+            sub_property_types.append(sub_property_type)
     if sub_property_types:
         if len(sub_property_types) == 1:
             return f"{property_type} of {sub_property_types[0]}"
@@ -832,6 +844,31 @@ def _flatten_schema_all_of(src_cfg: dict):
         src_cfg.update(sub_schema)
 
 
+def _get_property_description(prop_config: dict) -> str:
+    """Return accumulated property description.
+
+    Account for the following keys:
+    - top-level description key
+    - any description key present in each subitem under anyOf or allOf
+    """
+    description = prop_config.get("description", "")
+    oneOf = prop_config.get("oneOf", {})
+    anyOf = prop_config.get("anyOf", {})
+    for sub_item in chain(oneOf, anyOf):
+        if sub_item.get("description"):
+            prefix = " " if description else ""
+            if sub_item.get(DEPRECATED_KEY):
+                prefix += DEPRECATED_PREFIX
+            if description and description[-1] != ".":
+                prefix = f".{prefix}"
+            description += f"{prefix}{sub_item['description']}"
+    if bool(prop_config.get(DEPRECATED_KEY)):
+        description = _add_deprecation_msg(description)
+    if description:
+        description = " " + description
+    return description
+
+
 def _get_property_doc(schema: dict, defs: dict, prefix="    ") -> str:
     """Return restructured text describing the supported schema properties."""
     new_prefix = prefix + "    "
@@ -852,12 +889,7 @@ def _get_property_doc(schema: dict, defs: dict, prefix="    ") -> str:
             if prop_config.get("hidden") is True:
                 continue  # document nothing for this property
 
-            deprecated = bool(prop_config.get(DEPRECATED_KEY))
-            description = prop_config.get("description", "")
-            if deprecated:
-                description = _add_deprecation_msg(description)
-            if description:
-                description = " " + description
+            description = _get_property_description(prop_config)
 
             # Define prop_name and description for SCHEMA_PROPERTY_TMPL
             label = prop_config.get("label", prop_key)
diff --git a/cloudinit/config/schemas/schema-cloud-config-v1.json b/cloudinit/config/schemas/schema-cloud-config-v1.json
index e0ba98064..27b4e7e78 100644
--- a/cloudinit/config/schemas/schema-cloud-config-v1.json
+++ b/cloudinit/config/schemas/schema-cloud-config-v1.json
@@ -156,12 +156,12 @@
           "oneOf": [
             {
               "type": ["string", "null"],
-              "description": "Sudo rule to use or false. Absence of a sudo value or ``null`` will result in no sudo rules added for this user."
+              "description": "Sudo rule to use or ``null``. Absence of a sudo value or ``null`` will result in no sudo rules added for this user."
             },
             {
               "type": "boolean",
               "deprecated": true,
-              "description": "The value ``false`` will be deprecated in the future release. Use ``null`` or no ``sudo`` key instead."
+              "description": "The value ``false`` will be dropped after April 2027. Use ``null`` or no ``sudo`` key instead."
             }
           ]
         },
@@ -658,7 +658,6 @@
                   "description": "Specifies the partition table type, either ``mbr`` or ``gpt``. Default: ``mbr``."
                 },
                 "layout": {
-                  "type": ["string", "boolean", "array"],
                   "default": false,
                   "oneOf": [
                     {"type": "string", "enum": ["remove"]},
@@ -1394,7 +1393,7 @@
                     "type": "string",
                     "pattern": "^\\+?[0-9]+$",
                     "deprecated": true,
-                    "description": "Dropped after April 2027. Use ``now`` or integer type."
+                    "description": "Use of string for this value will be dropped after April 2027. Use ``now`` or integer type."
                 },
                 {"enum": ["now"]}
               ]
diff --git a/cloudinit/distros/__init__.py b/cloudinit/distros/__init__.py
index ffa41093f..4ceecbfb2 100644
--- a/cloudinit/distros/__init__.py
+++ b/cloudinit/distros/__init__.py
@@ -651,8 +651,16 @@ class Distro(persistence.CloudInitPickleMixin, metaclass=abc.ABCMeta):
             self.lock_passwd(name)
 
         # Configure sudo access
-        if "sudo" in kwargs and kwargs["sudo"] is not False:
-            self.write_sudo_rules(name, kwargs["sudo"])
+        if "sudo" in kwargs:
+            if kwargs["sudo"]:
+                self.write_sudo_rules(name, kwargs["sudo"])
+            elif kwargs["sudo"] is False:
+                LOG.warning(
+                    "DEPRECATED: The user %s has a 'sudo' config value of"
+                    " 'false' which is deprecated and will be removed in a"
+                    " future version. Use 'null' instead.",
+                    name,
+                )
 
         # Import SSH keys
         if "ssh_authorized_keys" in kwargs:
diff --git a/tests/unittests/config/test_cc_power_state_change.py b/tests/unittests/config/test_cc_power_state_change.py
index 5b9700026..8682b21fb 100644
--- a/tests/unittests/config/test_cc_power_state_change.py
+++ b/tests/unittests/config/test_cc_power_state_change.py
@@ -176,16 +176,18 @@ class TestPowerStateChangeSchema:
             (
                 {"power_state": {"mode": "halt", "delay": "5"}},
                 (
-                    "power_state.delay: DEPRECATED."
-                    " Dropped after April 2027. Use ``now`` or integer type."
+                    "power_state.delay: DEPRECATED. Use of string for this"
+                    " value will be dropped after April 2027. Use ``now`` or"
+                    " integer type."
                 ),
             ),
             ({"power_state": {"mode": "halt", "delay": "now"}}, None),
             (
                 {"power_state": {"mode": "halt", "delay": "+5"}},
                 (
-                    "power_state.delay: DEPRECATED."
-                    " Dropped after April 2027. Use ``now`` or integer type."
+                    "power_state.delay: DEPRECATED. Use of string for this"
+                    " value will be dropped after April 2027. Use ``now`` or"
+                    " integer type."
                 ),
             ),
             ({"power_state": {"mode": "halt", "delay": "+"}}, ""),
diff --git a/tests/unittests/config/test_cc_users_groups.py b/tests/unittests/config/test_cc_users_groups.py
index 12a00427a..92fdfc17e 100644
--- a/tests/unittests/config/test_cc_users_groups.py
+++ b/tests/unittests/config/test_cc_users_groups.py
@@ -448,7 +448,7 @@ class TestUsersGroupsSchema:
                     SchemaValidationError,
                     match=(
                         "deprecations: user.sudo: DEPRECATED. The value"
-                        " ``false`` will be deprecated in the future release."
+                        " ``false`` will be dropped after April 2027."
                         " Use ``null`` or no ``sudo`` key instead."
                     ),
                 ),
diff --git a/tests/unittests/config/test_schema.py b/tests/unittests/config/test_schema.py
index a1af2275a..086aa7ba8 100644
--- a/tests/unittests/config/test_schema.py
+++ b/tests/unittests/config/test_schema.py
@@ -1259,7 +1259,7 @@ class TestSchemaDocMarkdown:
                         },
                     },
                 },
-                "**prop1:** (UNDEFINED) <description>\n",
+                "**prop1:** (UNDEFINED) <description>. DEPRECATED. <deprecat",
             ),
             (
                 {
@@ -1280,7 +1280,7 @@ class TestSchemaDocMarkdown:
                         },
                     },
                 },
-                "**prop1:** (number)\n",
+                "**prop1:** (number) DEPRECATED. <deprecated_description>",
             ),
         ],
     )
```

blackboxsw commented on cloudinit/config/schema.py:
```diff
@@ -715,6 +716,23 @@ def _sort_property_order(value):
+
+
+def _collect_subschema_types(property_dict: dict, multi_key: str) -> List[str]:
+    property_types = [
+        subschema["type"]
+        for subschema in property_dict.get(multi_key, {})
+        if subschema.get("type") and not subschema.get("deprecated", False)
+    ]
```
I think we need to also take into account 'enum' keys too. Maybe something like the following:

```suggestion
    property_types = []                                                        
    for subschema in property_dict.get(multi_key, {}):                         
        if subschema.get(DEPRECATED_KEY):  # don't document deprecated types   
            continue                                                           
        if subschema.get("enum"):                                              
            property_types.extend(                                             
                [                                                              
                    f"``{_YAML_MAP.get(enum_value, enum_value)}``"             
                    for enum_value in subschema.get("enum", [])                
                ]                                                              
            )                                                                  
        elif subschema.get("type"):                                            
            property_types.append(subschema["type"])      
```

aciba90 replied:
Done.

blackboxsw commented on cloudinit/config/schemas/schema-cloud-config-v1.json:
```diff
@@ -153,8 +153,17 @@
+            {
+              "type": "boolean",
+              "deprecated": true,
+              "description": "The value ``false`` will be deprecated in the future release. Use ``null`` or no ``sudo`` key instead."
```
```suggestion
              "description": "The value ``false`` will be dropped after April 2027. Use ``null`` or no ``sudo`` key instead."
```

aciba90 replied:
done in c60e194b931ac88abd48c8887f6e77c0db185a64, thanks!

aciba90 commented:
During this PR, we have discovered two issues related to schema validation:

* https://warthogs.atlassian.net/browse/SC-1217
* https://warthogs.atlassian.net/browse/SC-1218

I have addressed the comments. This PR is ready to be reviewed, @blackboxsw.

blackboxsw commented on cloudinit/config/schema.py:
```diff
@@ -830,19 +859,25 @@ def _get_property_description(prop_config: dict) -> str:
     oneOf = prop_config.get("oneOf", {})
     anyOf = prop_config.get("anyOf", {})
-    descriptions = [] if not prop_descr else [prop_descr.rstrip(".")]
+    descriptions = []
+    deprecated_descriptions = []
+    if prop_descr:
+        prop_descr = prop_descr.rstrip(".")
+        if not prop_config.get(DEPRECATED_KEY):
+            descriptions.append(prop_descr)
+        else:
+            deprecated_descriptions.append(_add_deprecation_msg(prop_descr))
```
Thank you @aciba90 for unravelling my code additions to make this more legible in the future. This a good improvement

blackboxsw commented on cloudinit/config/schemas/schema-cloud-config-v1.json:
```diff
@@ -29,7 +29,8 @@
           "description": "Optional. Date on which the user's account will be disabled. Default: ``null``",
-          "type": "string"
+          "type": "string",
+          "format": "date"
```
nice hinting there for schema

blackboxsw reviewed (approved):
Nice work on the schema redacting for property docs. 
+1 doc diffs look well cleaned up and logic improvements on the description handling too.
